filebeat/reader/readfile/decoder.go

```diff
@@ -0,0 +1,169 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package readfile
+
+import (
+	"fmt"
+	"io"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+// decoderReader reads from an input file and converts its content into UTF-8.
+// It stores the size of the original bytes, so offset can be calculated later in
+// the line scanner.
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	bufferSize int
+	symlen     []uint8
+}
+
+func newDecoderReader(in io.Reader, codec encoding.Encoding, bufferSize int) *decoderReader {
+	return &decoderReader{
+		in:         in,
+		decoder:    codec.NewDecoder(),
+		buf:        streambuf.New(nil),
+		encodedBuf: streambuf.New(nil),
+		bufferSize: bufferSize,
+	}
+}
+
+func (r *decoderReader) read(buf []byte) (int, error) {
+	// TODO add to the beginning of buf and then start decoding
+	// instead of returning
+	if r.buf.Len() != 0 {
+		return r.copyToOut(buf)
+	}
+
+	b := make([]byte, r.bufferSize)
+	start := r.encodedBuf.Len()
+	n, err := r.in.Read(b[start:])
+	if err != nil {
+		return 0, err
+	}
+
+	if start > 0 {
+		enc, err := r.encodedBuf.Collect(start)
+		if err != nil {
+			return 0, err
+		}
+		r.encodedBuf.Reset()
+		b = append(enc, b[start:]...)
```

Comment:
more correct/efficient would be `b = append(enc, b[start:start+n]...)`. If `n < len(b[start:])`, then this append operation will add loads of '\0' to the end of b.